runtime/src/main/cpp/Memory.cpp

```diff
@@ -2252,6 +2253,56 @@ void FreezeSubgraph(ObjHeader* root) {
 #endif
 }
 
+void deepFrozenCopyTo(ObjHeader* to, ObjHeader* from, ObjHeaderMap* translationMap);
+
+void changeToCopyIfNeeded(ObjHeader** location, ObjHeaderMap* translationMap) {
+  auto* object = *location;
+  if (object == nullptr) return;
+  if (PermanentOrFrozen(object)) {
+    AddHeapRef(object);
+    return;
+  }
+  auto substitution = translationMap->find(object);
+  if (substitution != translationMap->end()) {
+    *location = substitution->second;
+    AddHeapRef(*location);
+  } else {
+    ObjHolder result;
+    auto* clone = AllocInstance(object->type_info(), result.slot());
+    SetHeapRef(location, clone);
+    deepFrozenCopyTo(clone, object, translationMap);
+  }
+}
+
+void deepFrozenCopyTo(ObjHeader* to, ObjHeader* from, ObjHeaderMap* translationMap) {
+   (*translationMap)[from] = to;
+   ::memcpy(to, from, objectSize(from));
```

Comment:
This behavior has a bug, if GC or exception is thrown in the middle of copying, as half-copied object is incorrect GC wise, need to come up with the better mechanism here.